### ISSUE TITLE
Return-Path: missing on some inbound emails, malformed on outbound virus bounce

### ISSUE THREAD

**Issue by cepheid666** [MailScanner/v5#418]
**OS:** `CentOS Linux release 7.7.1908 (Core)`
**MailScanner:** `MailScanner-5.1.3-2.noarch`
**sendmail:** `sendmail-8.14.7-5.el7.x86_64`

(1) In some (but not all!) inbound emails, the `Return-Path` header is missing.  I have not been able to find any specific pattern that causes this, and it also appears intermittent, i.e., mail from a given sender will be missing the header for some messages, but will have it for others.  I have no idea how to trace why some messages are missing the header, and why others are not.

**EDIT: (18 Dec 2019)** Root cause and suggested fix found, see https://github.com/MailScanner/v5/issues/418#issuecomment-566921189. **END EDIT**

(2) Outbound virus bounce notices (auto-generated by MailScanner) appear to have a malformed `Return-Path` header ... in the outbound messages in /var/spool/mqueue, I'm seeing this:

`H?P?Return-Path: <?g>`

A long Google search reveals very little but shows that [something like this was reported ages ago](https://mailscanner.mailscanner.narkive.com/Kyq2bawx/return-path-header-corrupt-in-virus-reports), so this may be a regression of that old bug.  (Possibly both (1) and (2) are related, since they're both mentioned in that linked thread.)

**EDIT: (18 Dec 2019)** Apparently the `<?g>` formatting is correct for modern sendmail. Sendmail will interpret this formatting appropriately in the queue file.  So this item (2) is a non-issue. **END EDIT**

Resolving (1) is my top priority, because those missing `Return-Path` headers mean that some downline programs (like SpamAssassin) cannot find the Envelope-From header (since it's not there!).  I can tell these programs to look for the X-MailScanner-From header, but that's a workaround, not a fix.

Note: MailScanner is set to _NOT_ use SpamAssassin (`Use SpamAssassin = No` in `/etc/MailScanner/MailScanner.conf`) on this config.  SpamAssassin is called downstream (sendmail delivers via procmail, which feeds into spamc with per-user preferences).

Any help is greatly appreciated.  Please let me know what debug/conf info I can provide, especially to figure out the intermittency of (1).

(I know v5.1.3-2 isn't the newest but I didn't see anything in the changelog that would indicate this particular issue would be resolved in the current version...)

Thanks!

**Comment by msapiro** [comment]
The Return-Path: header is added by the MTA upon "final delivery". It's content is the envelope sender - the address in the SMTP MAIL FROM command. See https://tools.ietf.org/html/rfc5321#section-4.4.

As far as I know, MailScanner doesn't manipulate/modify Return-Path: headers.. This is an MTA issue.

**Comment by cepheid666** [comment]
Looking at `/usr/share/MailScanner/perl/MailScanner/Sendmail.pm` and `SA.pm`, there is definitely manipulation of the `Return-Path` header going on.  In `Sendmail.pm` there is some rewriting of this header to deal with certain flags; in `SA.pm` this header is physically removed.  In my case, I have `Use SpamAssassin = No` so, in principle, the relevant code in `SA.pm` should never be called... though I don't know if this is actually the case.  The code in `Sendmail.pm` will be called, for sure, but I don't know if this could cause omission of the header, or just modifies it if it's there.

So MailScanner definitely does manipulate/modify, and in some cases removes, the `Return-Path` header.

When using sendmail as the MTA, MS modifies the daemon behavior such that two daemons handle delivery -- the first daemon queues the message, whereupon MailScanner scans/modifies it and deposits it into a second queue directory, after which the second daemon delivers it (using the delivery agent specified in sendmail.cf, which in my case is procmail).

The `Return-Path` header is added by the first daemon to the queue files that MailScanner looks at, but with sendmail macros (e.g., $g instead of the actual envelope address).  MailScanner then does whatever it does, and deposits the modified files into the second queue directory.

The second daemon then takes the queue files and substitutes all of the appropriate sendmail macros, but does not add additional headers.  Therefore, if MailScanner modified the headers, e.g., by removing the Return-Path header, the second sendmail daemon won't add it back in.

This is a MailScanner issue.  It's an MTA issue only insofar as how MailScanner interacts with the MTA, but it is definitely MailScanner doing the header removal.

I will note -- if I disable MailScanner, this header is NEVER missing.  This is the same behavior as noted in the linked thread from many ages ago.

Cheers.

**Comment by cepheid666** [comment]
Importantly, note that the RFC5321 text that you linked to specifies that:
> Here, final delivery means the message has left the SMTP environment.  Normally, this would mean it had been delivered to the destination user or an associated mail drop, but in some cases it may be further processed and transmitted by another mail system.

Although I could very well be wrong, my interpretation of that text is that the first sendmail daemon is the one doing "final delivery," into `/var/spool/mqueue.in` where MailScanner is looking for it.  Although this isn't actual delivery to the user, this is no longer within the SMTP environment and is therefore considered "final delivery" according to the RFC.

Then MailScanner does its thing, including modification or removal of the Return-Path header (see above comment).

Then, the second daemon is acting as a queue delivery agent and is no longer within the SMTP environment... from what I can tell, it just takes what's in the queue, replaces the appropriate macros, and sends it on to procmail. It doesn't add headers (including Received or Return-Path) because it's no longer handling SMTP.

**Comment by msapiro** [comment]
The code in SA.pm removes any Return-Path: headers and inserts a Return-Path: with content = the envelope sender of the message. The intent of this code is to replace the sendmail placeholder Return-Path header with a valid header. The bottom line is this only affects the message passed to SpamAssassin for analysis.

The code in Sendmail.pm manipulates flags in the sendmail queue entry. This could affect what the "second daemon" does with the Return-Path:, but I don't know enough about sendmail queue entries to comment.

**Comment by cepheid666** [comment]
> The bottom line is this only affects the message passed to SpamAssassin for analysis.

I hoped as much, so this is likely not the root cause.

> The code in Sendmail.pm manipulates flags in the sendmail queue entry. This could affect what the "second daemon" does with the Return-Path:, but I don't know enough about sendmail queue entries to comment.

This is why I wonder if it's related to issue (2) noted in the OP.  There appears to be some bug where, at least sometimes, the `Return-Path` for outgoing messages has an invalid `?g` where it should have a `$g`.  I'm wondering if MS is, on occasion, doing some invalid rewrite, and then the second sendmail daemon (that handles queue delivery) throws that line away because it's malformed.  This is a pure guess, of course.

AFAICT the second daemon does not add any headers, it just takes what's in the queue directory and tries to process it.

**Comment by cepheid666** [comment]
After reading `Sendmail.pm`, I'm wondering if the "second daemon" is actually non-existent, and instead MS directly invokes sendmail to deliver scanned/cleaned messages.  If that's the case then I don't know how Return-Path is supposed to be set in those messages.

I tested this by disabling MailScanner while leaving the sendmail daemon running, whereupon sendmail delivers mail into /var/spool/mqueue.in but nothing else happens (messages stay in the queue dir)... then I disabled the sendmail daemon and re-enabled MailScanner, and mail got delivered even though the sendmail daemon wasn't running... which suggests that MailScanner invokes sendmail directly.

I verified that each and every message in the incoming mail in /var/spool/mqueue.in _*had*_ the Return-Path header in it, but the delivered mail after MailScanner processing sometimes did not... so deletion is happening somewhere within MailScanner, but I don't know why or where.  Mails sent from various different domains all have the exact same header format, but some of them have Return-Path stripped and some do not.

Interestingly, gmail sent from desktop MUA (Apple Mail) or from mobile (iOS Mail) have Return-Path stripped before delivery (but, again, intact in mqueue.in!).  But gmail sent from the web interface has its header intact.  And mail from a different domain (comcast) from the exact same MUA also has header intact.  In all cases the Return-Path was there in the queue file (in mqueue.in), before MailScanner got to it.  So it's definitely MailScanner's influence, but I don't know why, where, or how.

Bottom line is I still don't know what the heck is going on, and why some messages have Return-Path while others don't.

Argh.

**Comment by shawniverson** [comment]
@cepheid666 This sounds like a bug to me, based on the details.  This detail you shared is pretty damning...

> I verified that each and every message in the incoming mail in /var/spool/mqueue.in had the Return-Path header in it, but the delivered mail after MailScanner processing sometimes did not... so deletion is happening somewhere within MailScanner, but I don't know why or where.

I'm thinking about how to troubleshoot this issue.  I would need to set up an environment with sendmail and have enough mail processing happening to trigger the bug(s).  For obvious reasons, I don't want to do that with my prod environments, and a test environment may not produce the right conditions since I would need to generate test emails.

@cepheid666 How open are you to the idea of some remote diagnosis?

**Comment by cepheid666** [comment]
> I would need to set up an environment with sendmail and have enough mail processing happening to trigger the bug(s).

The easiest/cheapest way to do this would be to set up an AWS image with CentOS 7.  I have such a server running and could basically mirror it, although this would exceed my "free tier" usage and therefore would cost a few bucks... if you have free tier usage available, I could provide you with my AMI although I'd need to sanitize user data out of it so either way I'd probably have to spin up a new machine out of my tier...

> @cepheid666 How open are you to the idea of some remote diagnosis?

I'm open though we'd need to figure out a way of doing this that would not compromise user data on the production machine.  Alternatively, if all you need is some debug output from MailScanner, you can send me new .pm files that produce that output and I can compile that output from a bunch of deliberate test emails (the same way I verified this bug in [this comment](https://github.com/MailScanner/v5/issues/418#issuecomment-554612690).

Or if you have some other ideas...

**Comment by shawniverson** [comment]
@cepheid666 Debug output is what I need, so I could definitely go that route and produce some pm files for you to generate some debug output surrounding code that manipulates the Return-Path header.

**Comment by cepheid666** [comment]
@shawniverson Sounds good.  When ready, please send pm files (with patch diffs, so I can be aware of the operative changes) and instructions for use/data collection.

**Comment by shawniverson** [comment]
@cepheid666 Will do.  I'll start getting this prepared for you.

**Comment by shawniverson** [comment]
@cepheid666

Commit
https://github.com/shawniverson/v5/commit/80773a5b42cbe07aec5895c1805f7af3ae86aa60

Install the new SA.pm and Sendmail.pm changes.

https://raw.githubusercontent.com/shawniverson/v5/112319debugreturnpath/common/usr/share/MailScanner/perl/MailScanner/SA.pm
https://github.com/shawniverson/v5/blob/112319debugreturnpath/common/usr/share/MailScanner/perl/MailScanner/Sendmail.pm

Restart MailScanner

Monitor mail log and capture relevant output on an affected message.

**Comment by shawniverson** [comment]
@cepheid666 

Additional Information and problem location

SA.pm is not called in your configuration, because it is disabled

This code does not get called.

```
  @SAheaders = grep !/^Return-Path\:/i, @SAheaders;
  unshift(@SAheaders, 'Return-Path: <' . $message->{from} . ">\n");
```

I believe this is triggering a missing Return-Path header in some cases, and in others, the malformed header due to the special placeholder.

**Comment by shawniverson** [comment]
Working on a solution when SA is not in the picture...

**Comment by shawniverson** [comment]
PR #425 posted, I am performing some regression testing.  Please test and report back @cepheid666 no need to use debug code.  I found the problem.

**Comment by cepheid666** [comment]
@shawniverson I see that you moved Return-Path addition out of SA.pm and into Sendmail.pm ... but what if a user is using a different mailer?  Exim via MS uses sendmail but what about for Zmailer and other supported options?  (Or, do they use a different Envelope From header, not Return-Path?)

Hopefully this resolves this ancient issue! =)

**Comment by shawniverson** [comment]
@cepheid666 Not applicable to other scanners, Return-Path is already properly formatted and present in the header.  I just confirmed postfix in regression testing. :wink:

**Comment by cepheid666** [comment]
Perfect. :-) I will test the code changes locally in the next couple of days and confirm that this is resolved on my system.

This fixes both issues? Ensures presence of inbound header and fixes malformed outbound header?

Cheers!

--- Amir
thumbed via iPhone

**Comment by shawniverson** [comment]
@cepheid666 Yes, both issues.  In the case of sendmail, no Return-Path can get written to the queue without being properly constructed.  It checks for Return-Path and rewrites it, or it adds it if it is not present.

**Comment by cepheid666** [comment]
@shawniverson I see you've closed this so I assume it's been sufficiently validated/verified.  Unfortunately I haven't had time to test this locally yet.  I'm sure it's fine, but I'll let you know in case I run into any problems later. So glad this ancient bug could get resolved! :-)

**Comment by shawniverson** [comment]
@cepheid666 Sounds good, I'll reopen if needed.

**Comment by cepheid666** [comment]
@shawniverson I finally had some downtime and spent a bunch of hours doing debugging and regression testing. Sorry that it took me so long.

Bottom line: I suggest reverting the changes from PR #425 (return to the prior code).  I'm including the full details in the next post, but since it will be a long post (to be thorough), I figured I'd put the suggestion here, up front, in case it gets lost in the text and/or you don't feel like reading the tome.

The good news is that I figured out the root cause of the issue, and have a suggested fix (as well as another suggestion for overall header handling).

The bad news is the new code causes two fatal errors, and unfortunately also doesn't fix the issue (because the root cause was different than I originally thought).  So, the new code should be reverted, and I'll submit a new fix for the true root cause.

See next post for the full gory details.

Cheers and thanks!

**Comment by cepheid666** [comment]
Full details from above:

1. The new code from PR #425 unfortunately causes two separate fatal errors.  I've identified a fix, but...
2. I discovered the reason for the original problem (missing Return-Path header in some emails), which has to do with a weird interaction between the (unfortunately common) case of having multiple Return-Path headers and how MS processes the Return-Path header flags (both the old code and the new).  _MS isn't deleting the Return-Path header at all_, but rather is causing it to be malformed and hence it gets dropped upon final delivery.  (Also, apparently the `<?g>` formatting in the sendmail queue file (qf) is normal/expected, not malformed, since it works OK upon delivery -- more details below.  So there is no need to force the format to be `<$from>`, as sendmail will apparently interpret the `<?g>` flag correctly.).
3. So, IMHO, the best fix is to revert to the old code (roll back PR #425), and then implement a new fix (see below) to change how MS deals with multiple Return-Path headers.  A more robust fix would change how MS deals with **all** headers, to preserve flags.

More details on each item are below.  In the immediate term, I recommend reverting the changes from PR #425.  I will submit a new PR for dealing with multiple Return-Path headers.

**1. New code -- errors and fixes**
From PR #425, in the modified [Sendmail.pm](https://github.com/MailScanner/v5/pull/425/commits/74528582ebb72b4a2f430f1a569bd9080177fe03#diff-b133f276ad07d1b5a54b6b71e0987dd4), new lines 471 & 479 include an extraneous newline. This generates a blank line in the post-processed sendmail queue file (`/var/spool/mail/qf*`), which breaks the qf format and causes sendmail to throw a fatal error:

```
Dec 17 19:52:52 mailhost sendmail[14051]: xBHJqmwK014038: SYSERR(root): readqf: ./qfxBHJqmwK014038: line 21: bad line ""
Dec 17 19:52:52 mailhost sendmail[14051]: xBHJqmwK014038: Losing ./qfxBHJqmwK014038: unrecognized line
```

This error occurs for every message.  The fix is to revert to the prior code (recommended), or to remove the final `\n` from both lines 471 & 479 (not recommended as this will still not address the root cause, see item 3 below).

The second problem is that all header lines subsequent to Return-Path (i.e., all headers from the original message) have no header prefix in the post-processed qf file.  They must start with either `H` or `H??` as they do in the incoming file (`/var/spool/mqueue.in/qf*`), but this `H` prefix is missing in the post-processed file (`/var/spool/mqueue/qf*`).  This also breaks the qf format and causes sendmail to throw a fatal error similar to:

```
Dec 17 20:09:57 mailhost sendmail[15061]: xBHJqmwK014038: SYSERR(root): SECURITY ALERT: extra or bogus data in queue file: Authentication-Results: [redacted].com;\n\tdkim=pass (2048-bit key) header.d=[redacted].com header.i=@[redacted].com header.b="TWQcnwPg"
Dec 17 20:09:57 mailhost sendmail[15061]: xBHJqmwK014038: Losing ./qfxBHJqmwK014038: bogus queue line
```

This is because the command that writes the prefix (`$h =~ s/^\S/H$&/;`) was moved inside two `if` blocks in the [new code](https://github.com/MailScanner/v5/pull/425/commits/74528582ebb72b4a2f430f1a569bd9080177fe03#diff-b133f276ad07d1b5a54b6b71e0987dd4) {see lines 472 & 480} and hence now only gets executed for Return-Path, and not for any other header.

The fix is to revert to the prior code (recommended), or to move line 472 outside the braces to line 474 just before the `push` command (not recommended as this will still not address the root cause, see item 3 below).

In doing all this testing, I also discovered that the new code has a semantic issue -- if Return-Path is missing and is added by MS, the [new code] (https://github.com/MailScanner/v5/pull/425/commits/74528582ebb72b4a2f430f1a569bd9080177fe03#diff-b133f276ad07d1b5a54b6b71e0987dd4) lines 476-482 will add this as the *last* header. This has significant potential to break DKIM for forwarded emails.  If adding the header, it must be added as the *first* header so as not to break DKIM.  But, in principle, this header should never be missing... sendmail will include it unless it's misconfigured (in which case that's not MailScanner's problem), and I've discovered why the header became missing in certain emails (see below).  So, in principle, MS should _never_ have to add this header, and **IMHO reverting to the old code is the best option.**


**2. Reason for original problem (missing Return-Path in some emails)**
Through this testing, I discovered that some email providers will add a Return-Path header on their end, before delivering to the final server.  If the final server uses sendmail (like mine), this causes TWO (or more) Return-Path headers in the final file -- the top one added by sendmail, and one or more below added by the sender provider and possible intermediary relays.

The most notable "offender" is gmail -- mail sent from an external MTA (e.g., Outlook, Apple Mail, etc.) through gmail SMTP servers will have a gmail-added Return-Path header after the (also gmail-added) `X-Google-Smtp-Source:` and `X-Received:` headers.  Mail sent through gmail's web interface will have those X- headers but NOT the additional Return-Path header.  This is true for all G-Suite email (I tested both with gmail directly and with two university accounts run through G-Suite).  At least a few other providers also do this, though I haven't compiled a list.

Without MailScanner, the excess Return-Path headers don't cause a problem for sendmail, because only the top-most one (added by sendmail) includes the `P` final-delivery flag (i.e., the header in the qf file looks like `H?P?Return-Path: <?g>`) while the one added by gmail has no flags (it looks like `H??Return-Path: <user@gmail>`).  Thus, under normal circumstances, sendmail would retain only the one with the `P` flag and drop all others.

Unfortunately, if MailScanner sees multiple Return-Path headers, _it retains the flags only from the most recently-read one, i.e., the bottom one_!  This is because [Sendmail.pm line 345](https://github.com/MailScanner/v5/blob/master/common/usr/share/MailScanner/perl/MailScanner/Sendmail.pm#L344) will overwrite the `returnpathflags` with each instance of Return-Path, so `returnpathflags` represents only the most recently-read (i.e., bottom-most) version of the header.  Because only the top-most version has the `P` flag, and the bottom one does not, this causes `returnpathflags` to OMIT the `P` flag.

Then, when MS re-writes the top-most Return-Path field, it writes it without the `P` flag, so the post-processed qf file has multiple Return-Path fields but NONE of them have the `P` flag. Sendmail will then drop all of these on the floor, and the final email will have no Return-Path header.

So the root cause is not that MS was deleting the header, but rather was causing it to be malformed, and sendmail dropped it.  The new code in PR #425 does not solve this root cause... but a fix is below.


**3. Fix, and discussion/question**
The fix to ensure that MS uses the correct flags from the top-most header, and not from any versions below, is to modify Sendmail.pm line 345 so that it runs only the first time it sees a Return-Path header, and not for any subsequent headers.  This should be as simple as setting a flag.

I'll submit a PR for this shortly.

See next post for discussion/question.

**Comment by cepheid666** [comment]
Following from above...

I am actually wondering, why does MS strip the `H` header prefix from any of the headers at all?  (This happens in [Sendmail.pm:ReadQf(), lines 339-341](https://github.com/MailScanner/v5/blob/master/common/usr/share/MailScanner/perl/MailScanner/Sendmail.pm#L339).) Sendmail may be locally configured to add some flags to certain headers, e.g., Reply-To or Content-Type or whatever, but MS will drop these flags on the floor and they're never written to the post-processed file (every post-processed header is prefixed with a simple `H`, no flags).  Line 342 temporarily stores the flags removed by line 341, but those flags are never saved anywhere except for the Return-Path, so any additional configured flags for other headers get lost.

Is there a reason that MS has to remove the header/flag prefix at all?  Is it doing processing on the headers that requires them to be stripped?  I see there is some processing of the Received lines to pull out IPs, though I'm not sure why (i.e., what the IP is used for), and in any case the header stored into `@{$message->{headers}` could (and IMHO should) still retain the prefix without affecting any of this processing.  Then, MS wouldn't have to re-add the header prefix and lose the flags.

An alternative fix would be to separately store the flags for each header, along with the header, but that's more complicated (since flags would need to be matched to header) and I think also not necessary (since I don't see why the prefix should be stripped to begin with).

Overall, the entire process of writing the post-processed Qf seems overly complex -- all the non-header metadata is rewritten as-is, but there is a lot of code devoted to writing the headers.  I'm sure there's a reason that I'm missing right now, but why doesn't MS just regurgitate the input Qf as-is, with the exception of adding the X-MailScanner headers (either at top or appended) ?  Is there any other processing done on the input headers that could cause them to be changed or removed?

In addition to the quick fix for the multiple Return-Path headers, my PR could try to clean up the entire Qf handling ... if MS doesn't actually need to do all that header processing, I could eliminate many of those header-handling functions and just copy the input Qf with appropriate prepending or appending of the X-MailScanner headers.

Let me know if I should do that...

Cheers and thanks!

**Comment by shawniverson** [comment]
@cepheid666 Thank you for the detailed feedback.

#425 is reverted.  It is clear from your post there are multiple issues, and since it doesn't fix the root problem anyway, it is unnecessary to try to attempt to fix.

I'll need to take a deeper dive, but the reason for the header manipulation is that MailScanner doesn't just process sendmail queue files.  The headers need to converge in a standard way from multiple different MTAs.  Therefore the @{$message->{headers}} should not have the prefixes.  However, they should be added back correctly in preparation for queuing after processing.  

H prefixes are added back here, but as you indicate, no logic is present for additional optional header prefixes, so we may need to address this issue...
https://github.com/MailScanner/v5/blob/master/common/usr/share/MailScanner/perl/MailScanner/Sendmail.pm#L463-L470

**Comment by shawniverson** [comment]
@cepheid666 any updates?  There is still a couple bugs here to address.

**Comment by cepheid666** [comment]
@shawniverson Thanks for the prior details and sorry for the silence -- it's the holiday season so schedule is wacky.  Fixing the Return-Path issue I think I can do with two lines around Sendmail.pm line 345, and will attempt to do so in the next week.

Addressing the header flags for ALL headers will take some more thought.  I'll probably open a new bug for it so that the discussion and subsequent PR will be more self-contained, since it is somewhat off the original topic of this bug.  I'll do that after I've given it a bit of thought and come up with a suggested resolution. =)  Hopefully in the next week if possible.

I think those are the only two bugs to address, unless I'm missing something.

Cheers and happy holidays!
